components/automate-ui/src/app/services/projects-filter/projects-filter.reducer.spec.ts

```diff
@@ -0,0 +1,549 @@
+import { map } from 'lodash/fp';
+
+import { EntityStatus } from 'app/entities/entities';
+import { ProjectConstants } from 'app/entities/projects/project.model';
+import {
+  ProjectsFilterState,
+  projectsFilterInitialState,
+  projectsFilterReducer,
+  ProjectsFilterOption
+} from './projects-filter.reducer';
+import { LoadOptions, LoadOptionsSuccess } from './projects-filter.actions';
+
+const UNASSIGNED_PROJECT_ID = ProjectConstants.UNASSIGNED_PROJECT_ID;
+const UNASSIGNED_PROJECT_LABEL = ProjectConstants.UNASSIGNED_PROJECT_LABEL;
+const ALL_RESOURCES_LABEL = ProjectConstants.ALL_RESOURCES_LABEL;
+const ALL_PROJECTS_LABEL = ProjectConstants.ALL_PROJECTS_LABEL;
+const MULTIPLE_PROJECTS_LABEL = ProjectConstants.MULTIPLE_PROJECTS_LABEL;
+const BADGE_GREY_STATE = false;
+const BADGE_BLUE_STATE = true;
+
+describe('projectsFilterReducer', () => {
+  const initialState: ProjectsFilterState = projectsFilterInitialState;
+
+  describe('LOAD_OPTIONS', () => {
+    const action = new LoadOptions();
+
+    it('sets status to loading', () => {
+      const { optionsLoadingStatus } = projectsFilterReducer(initialState, action);
+      expect(optionsLoadingStatus).toEqual(EntityStatus.loading);
+    });
+  });
+
+  describe('LOAD_OPTIONS_SUCCESS', () => {
+
+    it('sets status to success', () => {
+      const payload: ProjectsFilterOption[] = [
+        genProject('p1', false)
+      ];
+      const action = new LoadOptionsSuccess(payload);
+
+      const { optionsLoadingStatus } = projectsFilterReducer(initialState, action);
+
+      expect(optionsLoadingStatus).toEqual(EntityStatus.loadingSuccess);
+    });
+
+    it('without unassigned, sorts projects when storing them', () => {
+      const payload: ProjectsFilterOption[] = [
+        genProject('d-proj', false),
+        genProject('b-proj', false),
+        genProject('c-proj', false),
+        genProject('a-proj', false)
+      ];
+      const action = new LoadOptionsSuccess(payload);
+
+      const { options } = projectsFilterReducer(initialState, action);
+
+      expect(map('label', options)).toEqual(['a-proj', 'b-proj', 'c-proj', 'd-proj']);
+    });
+
+    it('with unassigned, sorts projects except unassigned at bottom when storing them', () => {
+      const payload: ProjectsFilterOption[] = [
+        genProject('d-proj', false),
+        genProject(UNASSIGNED_PROJECT_LABEL, true, UNASSIGNED_PROJECT_ID),
+        genProject('b-proj', false),
+        genProject('zz-proj', false),
+        genProject('a-proj', false)
+      ];
+      const action = new LoadOptionsSuccess(payload);
+
+      const { options } = projectsFilterReducer(initialState, action);
+
+      expect(map('value', options))
+        .toEqual(['a-proj', 'b-proj', 'd-proj', 'zz-proj', UNASSIGNED_PROJECT_ID]);
+    });
+
+    describe('with exactly one allowed project', () => {
```

Comment:
Starting here each test section checks the dropdown label, the adjacent caret, and the adjacent badge (value and color) showing the count of selected or available items.